web/src/components/forums/Editor/Toolbar/Embed/EmbedImage.jsx

```diff
@@ -0,0 +1,110 @@
+import React, {useState} from 'react';
+import {Box, Button, DialogTitle, DialogContent, Input, IconButton} from '@mui/material';
+import ImageIcon from '@mui/icons-material/Image';
+import {makeStyles} from '@mui/styles';
+import Popover from '@mui/material/Popover';
+
+export const useStyles = makeStyles((theme) => ({
+  button: {
+    margin: theme.spacing(0.5),
+  },
+}));
+
+export default function EmbedImage({embedImage}) {
+  const [clicked, setClicked] = useState(false);
+  const [imageURL, setImageURL] = useState('');
+  const [anchorEl, setAnchorEl] = React.useState(null);
+  const classes = useStyles();
+
+  const closeModal = () => {
+    setClicked(false);
+    setImageURL('');
+    setAnchorEl(null);
+  };
+
+  const openModal = (e) => {
+    setClicked(true);
+    setAnchorEl(e.currentTarget);
+  };
+
+  const handleEmbedImage = () => {
+    if (!imageURL) {
+      return;
+    }
+    embedImage(imageURL);
+    closeModal();
+  };
+
+  const uploadImage = (file) => {
+    const data = new FormData();
+    data.append('file', file);
+    axios.post(`/api/public/forums/image`, data)
+      .then((image_url) => {
+        // Wait for image url of uploaded image to be returned
+        return image_url.data;
+      })
+      .catch(standardErrorHandler(enqueueSnackbar));
+  };
+
+  const handleFileUpload = (e) => {
+    const file = e.target.files[0];
+    if (!file) {
+      return;
+    }
+    const reader = new FileReader();
+    reader.readAsDataURL(file);
+    reader.onloadend = () => {
+      const image_url = uploadImage(file);
+      embedImage(image_url);
+    };
+    closeModal();
+  };
```

Comment:
@dolf321 
A couple of things here lets start with uploadImage.

When you call an async function, and use `.then` you are giving it a callback once the promise is resolved, however the function itself will return before the promise has been resolved.

The return statement inside of the callback, returns from the callback not from the `uploadImage` function. And that will callback will be invoked asynchronously so when it executes your `uploadImage` function would have already returned.

instead you should do something like 
```javascript
const uploadImage = async (file) => {
  const data = new FormData();
  data.append('file', file);
  return await axios.post(`/api/public/forums/image`, data)
};
```

Now `uploadImage` itself is asynchronous, and when invoked it will make the http post request asynchronously and await it and return a promise, that will resolve to the image_url.

now within your handleFileUpload you can do:
```javascript
const handleFileUpload = (e) => {
  const file = e.target.files[0];
  if (!file) {
    return;
  }
  const reader = new FileReader();
  reader.readAsDataURL(file);
  reader.onloadend = () => {
    uploadImage(file)
      .then((image_url) => {
        embedImage(image_url);
      })
      .catch(standardErrorHandler(enqueueSnackbar));

  };
  closeModal();
};
```

now handleFileUpload is synchronous but you can pass a callback to the uploadImage callback which will then call the embedImage with the resolved image_url value.

Reply:
Thank you into putting the effort into a detailed explanation, I completely forgot that catch doesn't block the order of execution and instantly returns a promise unlike await.